### PR TITLE
Implement FixedFootDetector class

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
This PR implements the same logic used in the unicycle library to find the fixed foot given a list of planned contact.

Before reviewing the pr I have to:
- [x] write the test
- [x] Improve the documentation 

This code is meant to be used with the leggedodomerty class to get the robot base from a list of planned contact. 

cc @S-Dafarra and @prashanthr05

### CODE REVIEW

**Comment by S-Dafarra** [comment]
Initial picky comment. The name ``BaseDetector`` seems a bit off. It does not detect the base, as that is usually a choice. ``FixedFootDetector``?

**Comment by GiulioRomualdi** [comment]
I also don't like the name. I don't know if `FootFixedDetector` is the right choice since also the Smith trigger detector aims to find the stance foot. 

Probably @prashanthr05  has some idea

**Comment by prashanthr05** [comment]
> I also don't like the name. I don't know if `FootFixedDetector` is the right choice since also the Smith trigger detector aims to find the stance foot.
> 
> Probably @prashanthr05 has some idea

`SchmittTriggerDetector` just tries to detect if any given link is in contact or not. It does not make decision on the stance foot.

**Comment by prashanthr05** [comment]
Maybe we could eventually move these methods ithat currently exist in LeggedOdometry to determine the latest switched contact and the last active contact from a list of contacts using a switching time parameter. also into this class.

https://github.com/dic-iit/bipedal-locomotion-framework/blob/4ad31d8d62ffa0e3da5b7fbef77a945124c7bb4e/src/Estimators/src/LeggedOdometry.cpp#L336-L409

**Comment by GiulioRomualdi** [comment]
Documentation and test written. I also renamed the class as suggested by @S-Dafarra.
The PR is ready to be reviewed

File: src/Contacts/src/FixedFootDetector.cpp [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+                              [&](const auto& phase) -> bool {
+                                  return phase.beginTime <= m_currentTime
+                                         && m_currentTime <= phase.endTime;
+                              });
+
+    // take the last phase if not found
+    if (phase == m_contactPhaselist.cend())
+        phase = m_contactPhaselist.lastPhase();
+
+    // if the robot is in single support only once contact is enabled
```
Comment (S-Dafarra) [inline src/Contacts/src/FixedFootDetector.cpp:62]:
```suggestion
    // if the robot is in single support only one contact is enabled
```

File: src/Contacts/src/FixedFootDetector.cpp [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                contact.isActive = false;
+            } else
+            {
+                contact.isActive = true;
+                contact.pose = it->second->pose;
+            }
+        }
+    } else if (phase->activeContacts.size() == 2)
+    {
+        // Notice that here we do not analyze the case in which the phase is differnt from
```
Comment (S-Dafarra) [inline src/Contacts/src/FixedFootDetector.cpp:87]:
```suggestion
        // Notice that here we do not analyze the case in which the phase is different from
```

**Review by S-Dafarra — APPROVED** [review]

File: src/Contacts/include/BipedalLocomotion/ContactDetectors/ContactDetector.h [hunk 3]
```diff
@@ -8,105 +8,102 @@
...
+     * Enumerator used to determine the running state of the estimator
+     */
     enum class State
     {
-        NotInitialized, /**< The estimator is not initialized yet call FloatingBaseEstimator::initialze
-                           method to initialize it*/
-        Initialized,    /**< The estimator is initialized and ready to be used */
-        Running         /**< The estimator is running */
+        NotInitialized, /**< The estimator is not initialized yet call
+                           FloatingBaseEstimator::initialze method to initialize it*/
```
Comment (prashanthr05) [inline src/Contacts/include/BipedalLocomotion/ContactDetectors/ContactDetector.h:98]: Leftover from previous PRs 
```suggestion
                           ContactDetector::initialize method to initialize it*/
```

File: src/Contacts/include/BipedalLocomotion/ContactDetectors/FixedFootDetector.h [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * The aforementioned logic is described by the following diagram
+ * @code{.unparsed}
+ * time         0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17
+ * L            |+++|---|+++++++++++|---|+++++++++++|---|+++++++++++|---|+++++++++++|
+ * R            |+++++++++++|---|+++++++++++|---|+++++++++++|---|+++++++++++|---|+++|
+ * stance foot  |LLL|RRR|RRR|LLL|LLL|RRR|RRR|LLL|LLL|RRR|RRR|LLL|LLL|RRR|RRR|LLL|LLL|
+ * @endcode
+ * @warning The FixedFootDetector can handle only bipedal locomotion. Furthermore only one foot is
+ * considered as fixed foot
```
Comment (prashanthr05) [inline src/Contacts/include/BipedalLocomotion/ContactDetectors/FixedFootDetector.h:39]: Let's document this assumption more in detail maybe?

Saying that, the implementation assumes that the contact phase list only expects two contacts data and that at least one of the contact must be active at all times?

File: src/Contacts/src/FixedFootDetector.cpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        // firstPhase. Indeed in this case we should not update the contact states dictionary. (The
+        // contet of the dictionary is the one of the previous single support phase)
```
Comment (prashanthr05) [inline src/Contacts/src/FixedFootDetector.cpp:89]:
```suggestion
        // content of the dictionary is the one of the previous single support phase)
```

File: src/Contacts/src/FixedFootDetector.cpp [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+        // if the current phase is the first and there are at least 2 phases the active contact will
+        // be the one that in the next phase is active
```
Comment (prashanthr05) [inline src/Contacts/src/FixedFootDetector.cpp:92]:
```suggestion
        // be the one that is going to be active in the next phase
```

File: src/Contacts/tests/ContactDetectors/FixedFootDetectorUnitTest.cpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+#include <BipedalLocomotion/ParametersHandler/IParametersHandler.h>
+#include <BipedalLocomotion/ParametersHandler/StdImplementation.h>
+
+using namespace BipedalLocomotion::ParametersHandler;
+using namespace BipedalLocomotion::Contacts;
+
+struct FixedFootState
+{
+    EstimatedContact leftFoot;
+    EstimatedContact rightFoot;
```
Comment (prashanthr05) [inline src/Contacts/tests/ContactDetectors/FixedFootDetectorUnitTest.cpp:22]: Should this be a `PlannedContact` instead of an `EstimatedContact`?

**Review by prashanthr05 — COMMENTED** [review]
Minor Comments.

File: src/Contacts/tests/ContactDetectors/FixedFootDetectorUnitTest.cpp (hunk 7 above)
Comment (prashanthr05) [reply src/Contacts/tests/ContactDetectors/FixedFootDetectorUnitTest.cpp:22]: Nevermind, this is on the output end of the planner and input end of the estimator... so, `EstimatedContact` is fitting.

**Comment by GiulioRomualdi** [comment]
Once https://github.com/dic-iit/bipedal-locomotion-framework/pull/297 will be merged the tests shouldn't fail anymore

**Comment by GiulioRomualdi** [comment]
Hi @prashanthr05 and @S-Dafarra, in https://github.com/dic-iit/bipedal-locomotion-framework/pull/284/commits/113cfb9576f318882e339823775d5b18ba91cb51 I updated the logic used to generate the fixed foot. 

I also updated the documentation. Could you please check the PR?

**Review by prashanthr05 — APPROVED** [review]

**Pull request merged**